### ISSUE TITLE
Keyboard shortcuts in webview applications

### ISSUE THREAD

**Issue by bambooCZ** [nwjs/nw.js#6620]
Hello,

we are migrating to NWjs v0.30.x . We have noticed difference between 0.27.x in keyboard events in combination with webview.

We have keyboard shortcuts in our application that were relying on `keydown` event, like

```javascript
window.addEventListener("keydown", (evt) => {
    if (evt.ctrlKey) {
        switch (evt.key) {
            case "q": return nw.App.quit();
            case "n": return nw.Window.create(location.href);
            // etc...
        }
    }
});
```

But this has stopped to work, because the parent frame doesn't receive the `keydown` event when webview is focused.

**(IM)POSSIBLE WORKAROUNDS**

* `nw.Menu` - but we have undecorated window and thus this is unusable on Windows and Linux
    * We would require to be able to hide visual representation of the menu (maybe the `Alt` key to show it)
* `nw.Shortcut` - but it is system-wide and unregistereing them based on focus is somewhat unreliable
    * We would require to be able to register shortcut just for the application, not system-wide
* `chrome.commands` - the shortcut must start with a special key (like `Ctrl`, `Alt`, ...), so shortcuts like `F5` for reloading the webview are impossible
    * We would require to be able to register shortcut with any key combination, including a single key like `Esc`, `F5`, ...

Any proposals?

**Comment by panther7** [comment]
+1

**Comment by Christywl** [comment]
I can reproduce this issue on Linux/Windows with nwjs-sdk-v0.30.4.

**Comment by bambooCZ** [comment]
Can I hope for some resolution? Or is this condemned to death? Just asking if I should look for other possible solutions.

Another difficulty caused by this is that, we need to know if `Shift` or `Ctrl` is pressed when we receive for example `newwindow` event from webview, because further behavior depends on these modificators (new tab, or new window, ...). We have had a `keyboard` service in our app which held the state of keys and was based on `keydown` and `keyup` events. This has stopped to work also.

**Comment by GnorTech** [comment]
Will see it soon. thanks.

**Comment by GnorTech** [comment]
Could you provide reproducible code? This sample doesn't work for me with 0.27.x either:
index.html:
```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
</head>
<body style="width:800px; height:800px">
  <div style="width:800px; height:800px">
    <webview src="wv.html" partition="trusted" id="webview" style="display:block;width:700px; height:700px"></webview>
  </div>
</body>
</html>
```
wv.html:
```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
  </head>
  <body>
    <h1>webview</h1>
    <div id='output'></div>
    <script>
      window.addEventListener("keydown", (evt) => {
      if (evt.ctrlKey) {
      switch (evt.key) {
      case "q": document.getElementById('output').innerHTML = "ctrl-q pressed"; return true;
      }
      }
      });
  </body>
</html>
```
package.json:
```
{
  "name": "nw-demo",
    "main": "index.html",
    "webview": {
        "partitions": [
            {
                "name": "trusted",                                                                                                                                                                  "accessible_resources": ["<all_urls>"]
            }
        ]
    }
}
```

**Comment by bambooCZ** [comment]
@GnorTech Try this. You have misunderstood the problem. The events should be caught in the top frame not in the webview.

[keys.zip](https://github.com/nwjs/nw.js/files/2213217/keys.zip)

**Comment by GnorTech** [comment]
Thanks for the demo. Unfortunately I think the behavior in the new versions are correct. The DOM event is not supposed to bubble across DOM Window? See `bubble phase` in https://www.w3.org/TR/uievents/#event-flow

**Comment by bambooCZ** [comment]
OK I will not argue about that. But I am trying to implement keyboard shortcuts. What would you recommend? This was a way in the past. Now what?

**Comment by zahimich** [comment]
we need this bug fixed ASAP too. thanks.

**Comment by rogerwang** [comment]
I agree with https://github.com/nwjs/nw.js/issues/6620#issuecomment-406939043 that the current behavior is correct and we'd not change this.

As for the feature request of listening to key events in webview, could you inject a script into webview  and relay the event from there?

**Comment by panther7** [comment]
@rogerwang Yes, the current implementation does it.

**Comment by cipriancraciun** [comment]
As others have mentioned this issue is critical for those implementing anything touching `webview` elements...

And unfortunately the proposed workaround of injecting JS into the `webview` and using messages is not quite "safe", especially since `webview` elements are used to contain untrusted sources...

(Moreover given that for many releases this feature worked, and now it suddenly doesn't isn't very reassuring for the reliability and stability of the platform...)

**Comment by flddr** [comment]
Can anyone just guide me: i have a concept which could be build using nw.js. I have worked longer time ago with electron.

I want to have a parent window, which embeds a webview just as a child, to view it _like_ a movie. The webview should be safe, but it's just like a gimmick. The parent - the master - the worker - _that's the parent_, so every shortcut **must be handled by the parent**, not the webview.

What's the way right now?
